compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1211,4 +1206,18 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             err_ub!(DanglingIntPointer(offset, CheckInAllocMsg::InboundsTest)).into()
         })
     }
+
+    /// Check if a `src` ptr overlaps with a `dest` ptr.
+    #[inline(always)]
+    pub fn check_ptr_overlap(
+        src_id: AllocId,
+        src_offset: Size,
+        dest_id: AllocId,
+        dest_offset: Size,
+        size: Size,
+    ) -> bool {
+        src_id == dest_id
+            && ((src_offset <= dest_offset && src_offset + size > dest_offset)
+                || (dest_offset <= src_offset && dest_offset + size > src_offset))
```

Comment:
That's a very good improvement, good find.

My original suggestion was a bit unclear, I meant to only write the bool logic once and reuse it:

```suggestion
        let overlaps = |a, b| a <= b && b < a + size;
        src_id == dest_id
            && (overlaps(src_offset, dest_offset) || overlaps(dest_offset, src_offset))
```